### ISSUE TITLE
Unused vars not detected in destructuring assignment

### ISSUE THREAD

**Issue by rajivshah3** [eslint/eslint#11231]
**Tell us about your environment**
macOS Mojave 10.14.2 Beta

* **ESLint Version:** v4.19.1, also reproducible on v5.11.1
* **Node Version:** v8.14.0
* **npm Version:** v6.4.1 (I normally use Yarn v1.12.3 though)

**What parser (default, Babel-ESLint, etc.) are you using?** `babel-eslint`

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```yml
env:
    browser: true
    es6: true
    commonjs: true
    node: true
    mocha: true
    jest: true
parser: babel-eslint
parserOptions:
    ecmaVersion: 6
    ecmaFeatures:
        experimentalObjectRestSpread: true
        jsx: true
        modules: true
        arrowFunctions: true
        classes: true
        spread: true

    sourceType: module
extends:
    - airbnb
    - eslint:recommended
    - plugin:react/recommended
    - prettier/react
    - prettier
plugins:
    - babel
    - react
    - jsx-a11y
settings:
    import/resolver:
        node:
            extensions:
                - .js
                - .json
                - .css
            paths:
                - ./src/desktop/src
                - ./src/shared
                - ./src/mobile/src

rules:
    arrow-parens: [2, "always"]
    block-scoped-var: 2
    brace-style: [2, 1tbs, { "allowSingleLine": true }]
    camelcase: 2
    curly: [2, "all"]
    dot-location: [1, "property"]
    dot-notation: 2
    eol-last: 2
    eqeqeq: 2
    key-spacing: 2
    keyword-spacing: 2
    linebreak-style: [0, "unix"]
    no-alert: 2
    no-console: 1
    no-else-return: 1
    no-empty: [2, { "allowEmptyCatch": true }]
    no-eq-null: 2
    no-eval: 2
    no-extend-native: 2
    no-floating-decimal: 2
    no-implicit-coercion: [2, { "allow": ["!!"] }]
    no-implied-eval: 2
    no-lone-blocks: 2
    no-loop-func: 2
    no-multi-spaces: 2
    no-multi-str: 2
    no-native-reassign: 2
    no-new: 2
    no-new-func: 2
    no-new-wrappers: 2
    no-octal-escape: 2
    no-return-assign: 2
    no-script-url: 2
    no-self-compare: 2
    no-sequences: 2
    no-throw-literal: 2
    no-unmodified-loop-condition: 2
    no-unused-expressions: [2, { "allowShortCircuit": true, "allowTernary": true }]
    no-unused-vars: ["error"]
    no-useless-call: 2
    no-useless-concat: 2
    no-var: 2
    no-void: 2
    one-var: 2
    prefer-arrow-callback: 2
    prefer-const: 2
    quotes: [2, "single"]
    semi: 2
    yoda: 2
    import/extensions: 0
    import/no-extraneous-dependencies: 0
    import/prefer-default-export: 0
    space-in-parens: [2, "never"]
    # babel/no-invalid-this: 2
    jsx-a11y/alt-text: 0
    jsx-a11y/label-has-for: 0
    jsx-a11y/anchor-is-valid: 0
    jsx-a11y/no-autofocus: 0
    jsx-a11y/click-events-have-key-events: 0
    jsx-a11y/no-static-element-interactions: 0
    jsx-a11y/no-noninteractive-element-interactions: 0
    react/default-props-match-prop-types: 0
    react/forbid-prop-types: 0
    react/jsx-equals-spacing: 2
    react/jsx-filename-extension: [1, { "extensions": [".js", ".jsx"] }]
    react/jsx-first-prop-new-line: [1, "multiline"]
    react/jsx-indent: [2, 4]
    react/jsx-wrap-multilines:
        - 2
        -
            declaration: parens-new-line
            assignment: parens-new-line
            return: parens-new-line
            arrow: parens-new-line
            prop: ignore
    react/no-array-index-key: 0
    react/prefer-stateless-function: [1, {"ignorePureComponents": true}]
    react/require-default-props: 0

```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

Created a minimal example called `CustomTextInput-copy.js` with only the following code:

```js
const {
    label,
    theme,
    onChangeText,
    containerStyle,
    widget,
    onRef,
    testID,
    height,
    ...restProps
} = this.props;
```

I run ESLint using
```bash
yarn lint:mobile
```
This command triggers
```bash
./node_modules/.bin/eslint src/mobile/
```

The issue is specifically with `no-unused-vars: ["error"]`. With this alone, only `...restProps` is detected as unused. However, if I provide any `argsIgnorePattern`, the other unused variables are correctly detected (`label`, `theme`, etc). For example, `no-unused-vars: ["error", { "argsIgnorePattern": "" }]` or `no-unused-vars: ["error", { "argsIgnorePattern": "^_" }]` will correctly detect the unused variables.

**What did you expect to happen?**
All unused variables detected

**What actually happened? Please include the actual, raw output from ESLint.**
Only `restProps` is detected as unused when no `argsIgnorePattern` is specified
`10:8  error  'restProps' is assigned a value but never used  no-unused-vars`

**Are you willing to submit a pull request to fix this bug?**
I'm not sure what's causing this, so I can't submit a PR yet.

**Comment by platinumazure** [comment]
Hi @rajivshah3, thanks for the issue.

We have a [`ignoreRestSiblings` option](https://eslint.org/docs/rules/no-unused-vars#ignorerestsiblings) which is supposed to let users opt into the behavior you are seeing. Perhaps one of the configs you are extending is setting this property?

Try running ESLint with the `--print-config` option to see your combined config for the file you are linting. If you see `ignoreRestSiblings: true`, try setting `ignoreRestSiblings: false` in your configuration if you don't want that behavior.

We have special behavior when extending from a shareable config and specifying only a lint level (error/warning/off) in your extending config. In that case, the rule options from the extending config are preserved. (Personally, I wish we only did this when using a string configuration rather than an array configuration, since the array configuration is explicitly showing no other options, but that's just how things are right now.)

Hope this helps!

**Comment by rajivshah3** [comment]
Thanks for the reply @platinumazure . Here's the relevant config from `--print-config`:
```
"no-unused-vars": [
      "error",
      {
        "argsIgnorePattern": "^_"
      }
    ],
```
Interestingly, if I change the config to `no-unused-vars: ["error", { "ignoreRestSiblings": false }]` (removing `argsIgnorePattern` and adding `ignoreRestSiblings`), all unused variables are detected (as expected).

**Comment by rajivshah3** [comment]
> We have special behavior when extending from a shareable config and specifying only a lint level (error/warning/off) in your extending config. In that case, the rule options from the extending config are preserved. (Personally, I wish we only did this when using a string configuration rather than an array configuration, since the array configuration is explicitly showing no other options, but that's just how things are right now.)

Aha, I see what you mean now. When I do `print-config` with no other options except `"error"` for `no-unused-vars`, the `ignoreRestSiblings` is actually set to `true`:

```
"no-unused-vars": [
      "error",
      {
        "vars": "all",
        "args": "after-used",
        "ignoreRestSiblings": true
      }
    ],
```
Thanks! I'll close this as it's not a bug
